src/api/libcellml/parser.h

```diff
@@ -57,12 +57,39 @@ class LIBCELLML_EXPORT Parser: public Logger
      *
      * All existing issues will be removed before the input is parsed.
      *
+     * Returns a @c nullptr if the @p input is not a @c std::string representation
+     * of a CellML model.
+     *
      * @param input The string to parse into a model.
      *
      * @return The new @c ModelPtr deserialised from the input string.
      */
     ModelPtr parseModel(const std::string &input);
 
+    /**
+     * @brief Create and populate a new model from a @c std::string.
+     *
+     * Creates and populates a new model pointer by parsing CellML
+     * entities and attributes from CellML 1.0 and 1.1 format @p input @c std::string.
+     *
+     * Returns a @c nullptr if the @p input is not a @c std::string representation
+     * of a CellML model.
+     *
+     * All existing issues will be removed before the input is parsed.
+     *
+     * Set the @p renameNonSiUnits to @c true to rename common alternate spellings
+     * of SI units. Renaming non-SI units is @c false by default. The SI units
+     * that will get renamed if this parameter is true are:
+     *  * liter -> litre
+     *  * meter -> metre
+     *
+     * @param input The string to parse into a model.
+     * @param renameNonSiUnits Rename SI units with common alternate spellings.
+     *
+     * @return The new @c ModelPtr deserialised from the input string.
+     */
+    ModelPtr parse1XModel(const std::string &input, bool renameNonSiUnits = false);
```

Comment:
I think we need new API for CellML 1.X models because we want to know that the user is asking for an interpretation of their version 1.X model in version 2.0 terms.
I feel that the detection burden is very minor and not onerous, and I don't agree that libCellML should do that for you.

Reply:
Then, I am not liking it at all. libCellML handles CellML files while tools that use libCellML don't, and that's why they use libCellML. So, yes, I would expect libCellML to make their life as simple as possible when it comes to CellML, and it doesn't at this stage.

Tools that rely on libCellML should, using the same API, be able to parse any CellML contents, no matter whether it's CellML 1.X or 2.0. If libCellML parsed some CellML contents, then there should be a way to tell whether that contents was CellML 1.X or 2.0.

With the current API, those tools first need to detect whether they are dealing with a CellML 1.X or 2.0 file, and then call the appropriate API. To me, this is nonsense.

We are supposed to make libCellML's users' life easier...

Reply:
I'm starting to wonder if we actually want three different methods in the API depending on what the user wants to do.

Method 1 - parse a CellML 2 model: this is the current state and works well.

Method 2 - parse a CellML 1.x model and convert as much as possible to CellML 2: the user is knowingly asking libCellML to convert a CellML 1.x model into a CellML 2 model.

Method 3 - parse a CellML model: the user doesn't care what version CellML model they have and are prepared to do any follow up querying that might be needed to work out what is going on. This is really then just a convenience method that wraps the current `parseModel` with a fall-back to `parse1XModel`.

Reply:
I can say that for myself, I would use Method 3 all the time.  I would assume that it would then create reasonable CellML objects, which could be queried and possibly written back out to whatever CellML's current default level/version was.

I wouldn't use method 1 much, because no CellML 2 models exist :-)

I would use method 2 if method 3 didn't exist, and just assume that any CellML model I came across was 1.1, because that's the most likely scenario.

But really, I agree with @agarny that I don't want to worry about CellML levels and versions.  I just want to read in a file that the user told me is CellML, convert it to objects that I can query, translate those objects to Antimony objects, and be done.

Reply:
A "third" method might be the way to go. However, I would still have one method, but with a parameter specifying the "type" of inpuut, i.e. something like:
```c++
ModelPtr parseModel(const std::string &input, Type type);
```
where `Type` would be something like:
```c++
enum class Type
{
    AUTO,
    CELLML_1_X,
    CELLML_2_0
};
```
Also, in the case of `AUTO`, I wouldn't have it as a wrapper for the current `parseModel()` with a fall-back to `parse1XModel()`. Instead, I would have it as a proper parsing for CellML 1.X/2.0 models.

Reply:
You do need to start caring about such things when you are reading in a CellML 1.0 or CellML 1.1 document, you can't simply write out a CellML 2.0 document to the original location.  That would be disastrous.

Reply:
> You do need to start caring about such things when you are reading in a CellML 1.0 or CellML 1.1 document, you can't simply write out a CellML 2.0 document to the original location. That would be disastrous.

It's about to the libCellMLuser to decide what to do, not to libCellML. When it comes to OpenCOR, we want to read in any CellML file, be it CellML 1.X or CellML 2.0. When it comes to writing out a CellML file, it will depend on what we had to start with:
- New CellML file or CellML 2.0 file: we save the file as a CellML 2.0 file;
- CellML 1.X file: we warn the user that the file will be upgraded to CellML 2.0, and ask for confirmation to overwrite the current file or offer to save the file under a new name. libCellML has nothing to do with that.